src/rabbit_connection_tracking.erl

```diff
@@ -0,0 +1,320 @@
+%% at http://www.mozilla.org/MPL/
+%%
+%% Software distributed under the License is distributed on an "AS IS"
+%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+%% the License for the specific language governing rights and
+%% limitations under the License.
+%%
+%% The Original Code is RabbitMQ.
+%%
+%% The Initial Developer of the Original Code is GoPivotal, Inc.
+%% Copyright (c) 2007-2016 Pivotal Software, Inc.  All rights reserved.
+%%
+
+-module(rabbit_connection_tracking).
+
+%% Abstracts away how tracked connection records are stored
+%% and queried.
+%%
+%% See also:
+%%
+%%  * rabbit_connection_tracking_handler
+%%  * rabbit_reader
+%%  * rabbit_event
+
+-export([boot/0,
+         ensure_tracked_connections_table_for_node/1,
+         ensure_per_vhost_tracked_connections_table_for_node/1,
+         ensure_tracked_connections_table_for_this_node/0,
+         ensure_per_vhost_tracked_connections_table_for_this_node/0,
+         tracked_connection_table_name_for/1, tracked_connection_per_vhost_table_name_for/1,
+         clear_tracked_connections_table_for_this_node/0,
+         register_connection/1, unregister_connection/1,
+         list/0, list/1, list_on_node/1,
+         tracked_connection_from_connection_created/1,
+         tracked_connection_from_connection_state/1,
+         is_over_connection_limit/1, count_connections_in/1]).
+
+-include_lib("rabbit.hrl").
+
+-import(rabbit_misc, [pget/2]).
+
+%%
+%% API
+%%
+
+-spec boot() -> ok.
+
+%% Sets up and resets connection tracking tables for this
+%% node.
+boot() ->
+  ensure_tracked_connections_table_for_this_node(),
+  rabbit_log:info("Created a table for connection tracking on this node: ~p",
+    [tracked_connection_table_name_for(node())]),
+  ensure_per_vhost_tracked_connections_table_for_this_node(),
+  rabbit_log:info("Created a table for per-vhost connection counting on this node: ~p",
+    [tracked_connection_per_vhost_table_name_for(node())]),
+  clear_tracked_connections_table_for_this_node(),
+  ok.
+
+
+-spec ensure_tracked_connections_table_for_this_node() -> ok.
+
+ensure_tracked_connections_table_for_this_node() ->
+  ensure_tracked_connections_table_for_node(node()).
+
+
+-spec ensure_per_vhost_tracked_connections_table_for_this_node() -> ok.
+
+ensure_per_vhost_tracked_connections_table_for_this_node() ->
+  ensure_per_vhost_tracked_connections_table_for_node(node()).
+
+
+-spec ensure_tracked_connections_table_for_node(node()) -> ok.
+
+ensure_tracked_connections_table_for_node(Node) ->
+    TableName = tracked_connection_table_name_for(Node),
+    case mnesia:create_table(TableName, [{record_name, tracked_connection},
+                                       {attributes, [id, node, vhost, name,
+                                                     pid, protocol,
+                                                     peer_host, peer_port,
+                                                     username, connected_at]}]) of
```

Comment:
Any reason for not using `record_info(fields, ...)` for those two table creations?